File: rag/src/embed.rs
//! Embedding via fastembed (ONNX Runtime, no PyTorch). Maps the config's
//! HuggingFace-style model id to a fastembed model.

use std::path::PathBuf;

use anyhow::{Context, Result, bail};
use fastembed::{EmbeddingModel, InitOptions, TextEmbedding};

pub struct Embedder {
    model: TextEmbedding,
    pub dim: usize,
}

/// Supported model ids. Kept explicit so an unknown id is a hard error rather
/// than a silent downgrade that would build an index at the wrong dimension.
const SUPPORTED: &[(&str, EmbeddingModel)] = &[
    (
        "sentence-transformers/all-MiniLM-L6-v2",
        EmbeddingModel::AllMiniLML6V2,
    ),
    (
        "sentence-transformers/all-MiniLM-L12-v2",
        EmbeddingModel::AllMiniLML12V2,
    ),
    (
        "intfloat/multilingual-e5-small",
        EmbeddingModel::MultilingualE5Small,
    ),
    (
        "intfloat/multilingual-e5-base",
        EmbeddingModel::MultilingualE5Base,
    ),
    (
        "intfloat/multilingual-e5-large",
        EmbeddingModel::MultilingualE5Large,
    ),
    ("BAAI/bge-small-en-v1.5", EmbeddingModel::BGESmallENV15),
];

fn resolve_model(id: &str) -> Result<EmbeddingModel> {
    SUPPORTED
        .iter()
        .find(|(name, _)| *name == id)
        .map(|(_, m)| m.clone())
        .with_context(|| {
            let names: Vec<&str> = SUPPORTED.iter().map(|(n, _)| *n).collect();
            format!(
                "unknown embed model {id:?}. Supported: {}",
                names.join(", ")
            )
        })
}

/// Where downloaded ONNX models are cached. Defaults under ~/.engineering-os so
/// a run never litters the working directory. Override with EOS_MODEL_CACHE.
fn model_cache_dir() -> Result<PathBuf> {
    if let Ok(dir) = std::env::var("EOS_MODEL_CACHE") {
        return Ok(PathBuf::from(dir));
    }
    let home = std::env::var_os("HOME")
        .context("HOME is not set; set EOS_MODEL_CACHE to choose a model cache directory")?;
    Ok(PathBuf::from(home).join(".engineering-os").join("models"))
}

impl Embedder {
    pub fn new(model_id: &str) -> Result<Embedder> {
        let model = resolve_model(model_id)?;
        let mut model = TextEmbedding::try_new(
            InitOptions::new(model)
                .with_cache_dir(model_cache_dir()?)
                .with_show_download_progress(true),
        )?;
        let probe = model.embed(vec!["dimension probe"], None)?;
        let dim = probe.first().map(|v| v.len()).unwrap_or(0);
        if dim == 0 {
            bail!("embedding probe returned no vector; the model failed to produce embeddings");
        }
        Ok(Embedder { model, dim })
    }

    pub fn embed_batch(&mut self, texts: Vec<String>) -> Result<Vec<Vec<f32>>> {
        self.model.embed(texts, None)
    }

    pub fn embed_one(&mut self, text: &str) -> Result<Vec<f32>> {
        let mut out = self.model.embed(vec![text], None)?;
        out.pop()
            .context("embedding a single query returned no vector")
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn known_model_resolves() {
        assert!(resolve_model("sentence-transformers/all-MiniLM-L6-v2").is_ok());
        assert!(resolve_model("intfloat/multilingual-e5-small").is_ok());
    }

    #[test]
    fn unknown_model_is_an_error_not_a_downgrade() {
        let err = resolve_model("intfloat/multilingual-e5-larg").unwrap_err();
        assert!(err.to_string().contains("unknown embed model"));
    }
}

File: rag/src/embed.ts
// Embedding via transformers.js (ONNX, quantized). Whitelisted models only:
// an unknown id must be a hard error, never a silent downgrade that would
// build an index at the wrong dimension. e5-family models get the mandatory
// "passage: " / "query: " prefixes automatically.

import os from "node:os";
import path from "node:path";

export const DEFAULT_MODEL = "intfloat/multilingual-e5-small";

// id -> { dim, hub }: hub is the repo carrying ONNX weights for transformers.js.
const SUPPORTED: Record<string, { dim: number; hub: string }> = {
  "sentence-transformers/all-MiniLM-L6-v2": { dim: 384, hub: "Xenova/all-MiniLM-L6-v2" },
  "intfloat/multilingual-e5-small": { dim: 384, hub: "Xenova/multilingual-e5-small" },
  "intfloat/multilingual-e5-base": { dim: 768, hub: "Xenova/multilingual-e5-base" },
  "intfloat/multilingual-e5-large": { dim: 1024, hub: "Xenova/multilingual-e5-large" },
};

export function resolveModel(id: string): { dim: number; hub: string } {
  const entry = SUPPORTED[id];
  if (entry === undefined) {
    throw new Error(
      `unknown embed model ${JSON.stringify(id)}. Supported: ${Object.keys(SUPPORTED).join(", ")}`,
    );
  }
  return entry;
}

export function isE5(id: string): boolean {
  return id.includes("/multilingual-e5-") || id.includes("/e5-");
}

export function modelCacheDir(env: Record<string, string | undefined> = process.env): string {
  if (env.EOS_MODEL_CACHE !== undefined && env.EOS_MODEL_CACHE !== "") {
    return env.EOS_MODEL_CACHE;
  }
  return path.join(os.homedir(), ".engineering-os", "models");
}

type FeaturePipeline = (
  texts: string[],
  opts: { pooling: "mean"; normalize: boolean },
) => Promise<{ tolist(): number[][] }>;

export class Embedder {
  readonly dim: number;
  readonly modelId: string;
  private readonly pipe: FeaturePipeline;
  private readonly e5: boolean;

  private constructor(modelId: string, dim: number, pipe: FeaturePipeline) {
    this.modelId = modelId;
    this.dim = dim;
    this.pipe = pipe;
    this.e5 = isE5(modelId);
  }

  static async create(modelId: string, cacheDir?: string): Promise<Embedder> {
    const { dim, hub } = resolveModel(modelId);
    const { pipeline } = await import("@huggingface/transformers");
    const pipe = (await pipeline("feature-extraction", hub, {
      dtype: "q8",
      cache_dir: cacheDir ?? modelCacheDir(),
    })) as unknown as FeaturePipeline;
    const embedder = new Embedder(modelId, dim, pipe);
    const probe = await embedder.embedRaw(["dimension probe"]);
    const got = probe[0]?.length ?? 0;
    if (got !== dim) {
      throw new Error(`model ${modelId} produced dimension ${got}, expected ${dim}`);
    }
    return embedder;
  }

  private async embedRaw(texts: string[]): Promise<Float32Array[]> {
    if (texts.length === 0) return [];
    const out = await this.pipe(texts, { pooling: "mean", normalize: true });
    return out.tolist().map((v) => Float32Array.from(v));
  }

  async embedPassages(texts: string[]): Promise<Float32Array[]> {
    const prefixed = this.e5 ? texts.map((t) => `passage: ${t}`) : texts;
    return this.embedRaw(prefixed);
  }

  async embedQuery(text: string): Promise<Float32Array> {
    const prefixed = this.e5 ? `query: ${text}` : text;
    const out = await this.embedRaw([prefixed]);
    const vec = out[0];
    if (vec === undefined) throw new Error("embedding a query returned no vector");
    return vec;
  }
}
